### ISSUE TITLE
Issue about farmhash

### ISSUE THREAD

**Issue by p0d3r** [michaeldegroot/mysql-cache#38]
I don't know what's wrong with it but, every new environment i try to run my app, farmhash causes failure. There are a few error logs from my Amazon AWS instance:

```
4767 error Linux 4.4.44-39.55.amzn1.x86_64
4768 error argv "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/node" "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm" "--production" "rebuild"
4769 error node v6.10.0
4770 error npm  v3.10.10
4771 error code ELIFECYCLE
4772 error farmhash@1.2.1 install: `node-gyp rebuild`
4772 error Exit status 1
4773 error Failed at the farmhash@1.2.1 install script 'node-gyp rebuild'.
4773 error Make sure you have the latest version of node.js and npm installed.
4773 error If you do, this is most likely a problem with the farmhash package,
4773 error not with npm itself.
4773 error Tell the author that this fails on your system:
4773 error     node-gyp rebuild
4773 error You can get information on how to open an issue for this project with:
4773 error     npm bugs farmhash
4773 error Or if that isn't available, you can get their info via:
4773 error     npm owner ls farmhash
4773 error There is likely additional logging output above.
4774 verbose exit [ 1, true ]
```

```
npm ERR! You can get information on how to open an issue for this project with:
  npm ERR!     npm bugs farmhash
  npm ERR! Or if that isn't available, you can get their info via:
  npm ERR!     npm owner ls farmhash
  npm ERR! There is likely additional logging output above.
  
  npm ERR! Please include the following file with any support request:
  npm ERR!     /tmp/deployment/application/npm-debug.log
  Running npm install:  /opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm
  Setting npm config jobs to 1
  npm config jobs set to 1
  Running npm with --production flag
  Failed to run npm install. Snapshot logs for more details.
  Traceback (most recent call last):
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 695, in <module>
      main()
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 677, in main
      node_version_manager.run_npm_install(options.app_path)
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 136, in run_npm_install
      self.npm_install(bin_path, self.config_manager.get_container_config('app_staging_dir'))
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 180, in npm_install
      raise e
  subprocess.CalledProcessError: Command '['/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm', '--production', 'rebuild']' returned non-zero exit status 1 (ElasticBeanstalk::ExternalInvocationError)
caused by: + /opt/elasticbeanstalk/containerfiles/ebnode.py --action npm-install
  
  > xxhash@0.2.4 install /tmp/deployment/application/node_modules/xxhash
  > node-gyp rebuild
  
  make: Entering directory `/tmp/deployment/application/node_modules/xxhash/build'
    CXX(target) Release/obj.target/hash/src/hash.o
    CC(target) Release/obj.target/hash/deps/xxhash/xxhash.o
    SOLINK_MODULE(target) Release/obj.target/hash.node
    COPY Release/hash.node
  make: Leaving directory `/tmp/deployment/application/node_modules/xxhash/build'
  api@0.0.0 /tmp/deployment/application
  â””â”€â”€ xxhash@0.2.4 
  
  
  > farmhash@1.2.1 install /tmp/deployment/application/node_modules/farmhash
  > node-gyp rebuild
  
  gyp ERR! clean error 
  gyp ERR! stack Error: EACCES: permission denied, rmdir 'build'
  gyp ERR! stack     at Error (native)
  gyp ERR! System Linux 4.4.44-39.55.amzn1.x86_64
  gyp ERR! command "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/node" "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
  gyp ERR! cwd /tmp/deployment/application/node_modules/farmhash
  gyp ERR! node -v v6.10.0
  gyp ERR! node-gyp -v v3.4.0
  gyp ERR! not ok 
  
  npm ERR! Linux 4.4.44-39.55.amzn1.x86_64
  npm ERR! argv "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/node" "/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm" "--production" "rebuild"
  npm ERR! node v6.10.0
  npm ERR! npm  v3.10.10
  npm ERR! code ELIFECYCLE
  npm ERR! farmhash@1.2.1 install: `node-gyp rebuild`
  npm ERR! Exit status 1
  npm ERR! 
  npm ERR! Failed at the farmhash@1.2.1 install script 'node-gyp rebuild'.
  npm ERR! Make sure you have the latest version of node.js and npm installed.
  npm ERR! If you do, this is most likely a problem with the farmhash package,
  npm ERR! not with npm itself.
  npm ERR! Tell the author that this fails on your system:
  npm ERR!     node-gyp rebuild
  npm ERR! You can get information on how to open an issue for this project with:
  npm ERR!     npm bugs farmhash
  npm ERR! Or if that isn't available, you can get their info via:
  npm ERR!     npm owner ls farmhash
  npm ERR! There is likely additional logging output above.
  
  npm ERR! Please include the following file with any support request:
  npm ERR!     /tmp/deployment/application/npm-debug.log
  Running npm install:  /opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm
  Setting npm config jobs to 1
  npm config jobs set to 1
  Running npm with --production flag
  Failed to run npm install. Snapshot logs for more details.
  Traceback (most recent call last):
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 695, in <module>
      main()
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 677, in main
      node_version_manager.run_npm_install(options.app_path)
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 136, in run_npm_install
      self.npm_install(bin_path, self.config_manager.get_container_config('app_staging_dir'))
    File "/opt/elasticbeanstalk/containerfiles/ebnode.py", line 180, in npm_install
      raise e
  subprocess.CalledProcessError: Command '['/opt/elasticbeanstalk/node-install/node-v6.10.0-linux-x64/bin/npm', '--production', 'rebuild']' returned non-zero exit status 1 (Executor::NonZeroExitStatus)


[2017-03-16T07:28:02.051Z] INFO  [19762] - [Application update app-72065-170316_102706@19/AppDeployStage0/AppDeployPreHook/50npm.sh] : Activity failed.
[2017-03-16T07:28:02.051Z] INFO  [19762] - [Application update app-72065-170316_102706@19/AppDeployStage0/AppDeployPreHook] : Activity failed.
[2017-03-16T07:28:02.051Z] INFO  [19762] - [Application update app-72065-170316_102706@19/AppDeployStage0] : Activity failed.
[2017-03-16T07:28:02.051Z] INFO  [19762] - [Application update app-72065-170316_102706@19] : Completed activity. Result:
  Application update - Command CMD-AppDeploy failed
[2017-03-16T07:32:41.207Z] INFO  [20189] - [CMD-TailLogs] : Starting activity...
[2017-03-16T07:32:41.207Z] INFO  [20189] - [CMD-TailLogs/AddonsBefore] : Starting activity...
[2017-03-16T07:32:41.207Z] INFO  [20189] - [CMD-TailLogs/AddonsBefore] : Completed activity.
[2017-03-16T07:32:41.207Z] INFO  [20189] - [CMD-TailLogs/TailLogs] : Starting activity...
[2017-03-16T07:32:41.207Z] INFO  [20189] - [CMD-TailLogs/TailLogs/TailLogs] : Starting activity...
```

**Comment by michaeldegroot** [comment]
Have you tried https://github.com/michaeldegroot/mysql-cache#troubleshooting ?

**Comment by michaeldegroot** [comment]
I had a issue with travis with farmhash this was the fix:

```
sudo apt-get install python-software-properties
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt-get install gcc-5 g++-5
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 80 --slave /usr/bin/g++ g++ /usr/bin/g++-5
sudo update-alternatives --config gcc

// (choose gcc-5 from the list)
```

**Comment by michaeldegroot** [comment]
Also I see a 

` stack Error: EACCES: permission denied, rmdir 'build'`

Check permissions just to be sure :)

**Comment by p0d3r** [comment]
It's okay, i was able to fix issue on my windows development machine by installing npm build tools for windows but since i cannot access to amazon instances which are behind a load balancer, it cannot be fixed. i am currently using 2.1.0

**Comment by michaeldegroot** [comment]
That's unfortunate, You cannot update your GCC on this amazon instance? maybe I should make the new hashing options OPTIONAL and you access it with npm install and selecting it in the mysql-cache options?

**Comment by michaeldegroot** [comment]
This problem will be solved in the upcoming version 2.1.2 which will probably roll out today if the unit test gods like me today

> Modules like farmhash, xxhash and mmap-object will now be automatically installed. But only if the module is about to be used, so these packages are now optional and will not cause any problems for people that can't compile them.
